### PR TITLE
feat: @discordjs/brokers

### PR DESCRIPTION
**Pull request by didinele**
**Please describe the changes this PR makes and why it should be merged:**
message brokers batchest

4th DD lib here we go

### CODE REVIEW

File: packages/brokers/.cliff-jumperrc.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+	"name": "proxy",
+	"org": "discordjs",
+	"packagePath": "packages/proxy"
```
Comment (iCrawl) [inline packages/brokers/.cliff-jumperrc.json:4]: Ha.

Comment (didinele) [reply packages/brokers/.cliff-jumperrc.json:4]: fuck

File: packages/brokers/package.json [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	],
+	"license": "Apache-2.0",
+	"keywords": [
+		"discord",
+		"api",
+		"rest",
+		"proxy",
+		"discordapp",
+		"discordjs"
+	],
```
Comment (almeidx) [inline packages/brokers/package.json:47]: Doesn't look like these were changed. Should they?

Comment (didinele) [reply packages/brokers/package.json:47]: yup. i forgot

**Comment by didinele** [comment]
ready for review since unit testing any more of this would be hell without proper redis mocks lmao

**Comment by kyranet** [comment]
You have merge conflicts, @didinele

**Comment by didinele** [comment]
> You have merge conflicts, @didinele

fixed. moved to neon and back to tsup

**Comment by vladfrangu** [comment]
What can I say except..rebaseee

File: packages/brokers/src/brokers/redis/BaseRedis.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		this.streamReadClient = options.redisClient.duplicate();
+	}
+
+	public async subscribe(group: string, events: (keyof TEvents)[]): Promise<void> {
+		await Promise.all(
+			// eslint-disable-next-line consistent-return
+			events.map(async (event) => {
+				this.subscribedEvents.add(event as string);
+				try {
+					return await this.options.redisClient.xgroup('CREATE', event as string, group, 0, 'MKSTREAM');
```
Comment (vladfrangu) [inline packages/brokers/src/brokers/redis/BaseRedis.ts:51]: Is the return necessary?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/brokers/src/brokers/redis/BaseRedis.ts (hunk 3 above)
Comment (didinele) [reply packages/brokers/src/brokers/redis/BaseRedis.ts:51]: I suppose not. It just seemed intuitive to me given it's inside a `.map` call fed to `Promise.all`, but the behavior would indeed be the same without it.

File: .github/labeler.yml [hunk 4]
```diff
@@ -6,6 +6,11 @@ apps:website:
   - apps/website/**/*
 
 packages:builders:
+'packages:brokers':
+  - packages/brokers/*
+  - packages/brokers/**/*
+
+'packages:builders':
```
Comment (imranbarbhuiya) [inline .github/labeler.yml:13]:
```suggestion
packages:brokers:
  - packages/brokers/*
  - packages/brokers/**/*

packages:builders:
```

File: packages/brokers/package.json [hunk 5]
```diff
@@ -0,0 +1,85 @@
+{
+	"name": "@discordjs/brokers",
+	"version": "0.1.0-dev",
+	"description": "Powerful set of message brokers",
+	"scripts": {
+		"test": "vitest run",
+		"build": "tsup",
+		"lint": "prettier --check . && TIMING=1 eslint src __tests__ --ext mjs,js,ts",
```
Comment (imranbarbhuiya) [inline packages/brokers/package.json:8]: add changes from https://github.com/discordjs/discord.js/commit/e8f63617a231c028591a5cdd008bd0bb5f5f43cb

**Review by imranbarbhuiya — CHANGES_REQUESTED** [review]

File: packages/brokers/README.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+- [Discord API Discord server](https://discord.gg/discord-api)
+- [GitHub](https://github.com/discordjs/discord.js/tree/main/packages/brokers)
+- [npm](https://www.npmjs.com/package/@discordjs/brokers)
+- [Related libraries](https://discord.com/developers/docs/topics/community-resources#libraries)
+
+## Contributing
+
+Before creating an issue, please ensure that it hasn't already been reported/suggested, and double-check the
+[documentation](https://discord.js.org/#/docs/brokers).  
+See [the contribution guide](https://github.com/discordjs/discord.js/blob/main/.github/CONTRIBUTING.md) if you'd like to submit a PR.
```
Comment (iCrawl) [inline packages/brokers/README.md:102]: Check the other readmes for link mapping

File: packages/brokers/package.json [hunk 7]
```diff
@@ -0,0 +1,87 @@
+{
+	"name": "@discordjs/brokers",
+	"version": "0.1.0-dev",
```
Comment (iCrawl) [inline packages/brokers/package.json:3]:
```suggestion
	"version": "0.1.0",
```

File: packages/brokers/package.json [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+	"description": "Powerful set of message brokers",
+	"scripts": {
+		"test": "vitest run",
+		"build": "tsup",
+		"lint": "prettier --check . && cross-env TIMING=1 eslint src __tests__ --ext .mjs,.js,.ts --format=pretty",
+		"format": "prettier --write . && cross-env TIMING=1 eslint src __tests__ --ext .mjs,.js,.ts --fix --format=pretty",
+		"fmt": "yarn format",
+		"docs": "downlevel-dts dist docs/dist --to=3.7 && api-extractor run --local",
```
Comment (iCrawl) [inline packages/brokers/package.json:11]:
```suggestion
		"docs": "api-extractor run --local",
```

File: packages/brokers/package.json [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+		"cross-env": "^7.0.3",
+		"downlevel-dts": "^0.10.1",
+		"eslint": "^8.25.0",
+		"eslint-config-neon": "^0.1.38",
+		"eslint-formatter-pretty": "^4.1.0",
+		"prettier": "^2.7.1",
+		"rollup-plugin-typescript2": "^0.33.0",
+		"tsup": "^6.2.3",
+		"typescript": "^4.8.4",
+		"vitest": "^0.22.1"
```
Comment (iCrawl) [inline packages/brokers/package.json:79]:
```suggestion
		"@discordjs/docgen": "workspace:^",
		"@favware/cliff-jumper": "^1.8.8",
		"@microsoft/api-extractor": "^7.32.1",
		"@types/node": "^16.11.52",
		"@vitest/coverage-c8": "^0.22.1",
		"cross-env": "^7.0.3",
		"eslint": "^8.25.0",
		"eslint-config-neon": "^0.1.38",
		"eslint-formatter-pretty": "^4.1.0",
		"prettier": "^2.7.1",
		"tsup": "^6.2.3",
		"typescript": "^4.8.4",
		"vitest": "^0.22.1"
```

File: packages/brokers/api-extractor.json [hunk 10]
```diff
@@ -0,0 +1,376 @@
+/**
```
Comment (almeidx) [inline packages/brokers/api-extractor.json:1]: This file should just be 
```json
{
	"extends": "../../api-extractor.json"
}
```

**Review by iCrawl — APPROVED** [review]

File: packages/brokers/cliff.toml [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+	{ message = "^style", group = "Styling"},
+	{ message = "^test", group = "Testing"},
+	{ message = "^chore", skip = true},
+	{ message = "^ci", skip = true},
+	{ message = "^build", skip = true},
+	{ body = ".*security", group = "Security"},
+]
+filter_commits = true
+tag_pattern = "@discordjs\\/brokers@.*"
+skip_tags = "v[0-9]*|11|12"
```
Comment (almeidx) [inline packages/brokers/cliff.toml:61]: According to the template, this is how it should be
```suggestion
tag_pattern = "@discordjs/brokers@[0-9]*"
```

File: packages/brokers/README.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+[website]: https://discord.js.org/
+[website-source]: https://github.com/discordjs/discord.js/tree/main/apps/website
+[documentation]: https://discord.js.org/#/docs/ws
+[guide]: https://discordjs.guide/
+[guide-source]: https://github.com/discordjs/guide
+[guide-update]: https://discordjs.guide/additional-info/changes-in-v14.html
+[discord]: https://discord.gg/djs
+[discord-api]: https://discord.gg/discord-api
+[source]: https://github.com/discordjs/discord.js/tree/main/packages/ws
+[npm]: https://www.npmjs.com/package/@discordjs/ws
```
Comment (iCrawl) [inline packages/brokers/README.md:118]:
```suggestion
[source]: https://github.com/discordjs/discord.js/tree/main/packages/brokers
[npm]: https://www.npmjs.com/package/@discordjs/brokers
```

**Pull request merged**